### PR TITLE
feat(cluster): add slots blocking when we finalize migration

### PR DESCRIPTION
**Pull request by BorysTheDev**

### CODE REVIEW

File: src/server/cluster/cluster_config.h
Comment (adiholden) [inline src/server/cluster/cluster_config.h:82]: ;

Comment (BorysTheDev) [reply src/server/cluster/cluster_config.h:82]: it's not for review, just for me)

File: src/server/cluster/cluster_family.cc
Comment (dranikpg) [inline src/server/cluster/cluster_family.cc:724]: 1. server_family_service_proactor_pool should be the same as shard_set
2. Why do need two dispatches? Don't you set just thread local stuff?
3. SetMigratedSlots doesn't await anything, so it can be a RunBrief

File: src/server/cluster/cluster_family.cc
Comment (dranikpg) [inline src/server/cluster/cluster_family.cc:730]: same here

File: src/server/main_service.cc
Comment (dranikpg) [inline src/server/main_service.cc:1289]: there is another check below

File: src/server/cluster/cluster_config.h
Comment (dranikpg) [inline src/server/cluster/cluster_config.h:114]: You don't use it and nit: it's nicer to pass absl::Span<const SlotRange> 😅

File: src/server/main_service.cc
Comment (dranikpg) [inline src/server/main_service.cc:882]: this is an unfortunate place to block, verify function should be immediate... Can we block somehow from DispatchCommand?

Comment (dranikpg) [reply src/server/main_service.cc:882]: TODO(myself): Add some comments or DCHECKs about this

The reason for is that Verify commands are called when either collecting multi commands or preparing squashing  and sometimes they're not on the same code path as the executed command

File: src/server/cluster/cluster_family.cc
Comment (dranikpg) [inline src/server/cluster/cluster_family.cc:751]: DispatchTracker needs to run on the same non-preemptive operation when we change status. Scenario:

1. The dispatch above sets `SetMigratedSlots()`, so since this point in time clients can block on `AwaitIfMigrationFinalization`
2. Now some clients are blocked
3. You use the `DispatchTracker` and it waits for the clients blocked on `AwaitIfMigrationFinalization` (it can't distinguish them from a real operation)

This is the definition of "shoot oneself in the foot" 🙂 You should re-think the order of operations, because I assume we want to block operations first, then finalize, then unblock them. Currently the DispatchTracker doesn't any useful work

File: src/server/main_service.cc
Comment (dranikpg) [reply src/server/main_service.cc:882]: Actually it's very hard now 😅 😅 😅 

Scenario:
```
MULTI
SET a 1 -> ok stored
SET b 1 -> ok stored
SET C 1 -> ok stored 
-- finalize-slot-migration arrives --
EXEC -> we need to check a b c here, not above
```

Will come up with ideas tomorrow

**Review by dranikpg — COMMENTED** [review]
We should run it on some tests test first, because it looks to me that the current code doesn't really do what we expect it to 😅

File: src/server/main_service.cc
Comment (dranikpg) [reply src/server/main_service.cc:882]: I suggest using basic client pause for now. We have enough problems in the existing migration code, so why should we strive for a perfect solution here. To reduce the blocking time we can lower the replication gap ahead

File: src/server/cluster/cluster_family.cc
Comment (dranikpg) [inline src/server/cluster/cluster_family.cc:720]: You can't keep local references to something that you hand off, imagine:
1. polling fiber checks state -> false, so it sleeps
2. you finish migration -> scope aborts
3. polling fiber wakes up and reads from in invalid scope

Comment (dranikpg) [reply src/server/cluster/cluster_family.cc:720]: Also you don't need it to be atomic because we assume both fibers stay on the same thread, if we don't we have more problems than just an atomic 🙂

Comment (dranikpg) [reply src/server/cluster/cluster_family.cc:720]: What comes to mind is either retuning fiber ref from Pause or using shared_ptr<bool> for polling, maybe there are more options

**Review by dranikpg — COMMENTED** [review]
Sorry for the delay. Almost here, last small bug 🙂

File: src/server/server_family.cc
Comment (dranikpg) [inline src/server/server_family.cc:445]: Ah.. autoformatting 😆

**Review by dranikpg — APPROVED** [review]

File: src/server/cluster/cluster_family.cc
Comment (adiholden) [inline src/server/cluster/cluster_family.cc:719]: Do we have some timeout for the migration finalize ?

**Pull request merged**
